text/0580-destroyables.md

```diff
@@ -0,0 +1,243 @@
+associate them with a parent destroyable.
+
+```js
+class TimeoutManager {
+  constructor(parent, fn, timeout = 1000) {
+    let timeoutId = setTimeout(fn, timeout);
+
+    associateDestroyableChild(parent, this);
+    registerDestructor(this, () => clearTimeout(timeoutId))
+  }
+}
+
+class MyComponent extends Component {
+  manager = new TimeoutManager(this, () => console.log('hello'));
+}
+```
+
+## Motivation
+
+Ember manages the lifecycles and lifetimes of many built in constructs, such as
+components, and does so in a hierarchical way - when a parent component is
+destroyed, all of its children are destroyed as well. This is a well established
+software pattern that is useful for many applications, and there are a variety
+of libraries, such as [ember-lifeline](https://github.com/ember-lifeline/ember-lifeline)
+and [ember-concurrency](https://github.com/machty/ember-concurrency), that would
+benefit from having a way to extend this hierarchy, adding their own "children"
+that are cleaned up whenever their parents are removed.
+
+Historically, Ember has exposed this cleanup lifecycle via _hooks_, such as the
+`willDestroy` hook on components. However, methods like these have a number of
+downsides:
+
+1. Since they are named, they can have collisions with other properties. This is
+   historically what led to the actions hash on classic components, in order to
+   avoid collisions between actions named `destroy` and the `destroy` lifecyle
+   hook.
+2. On a related note, relying on property names means that all framework classes
+   _must_ implement the `willDestroy` function (or another name), making it very
+   difficult to change APIs in the future.
+3. Methods are difficult for _libraries_ to instrument. For instance,
+   `ember-concurrency` currently replaces the `willDestroy` method on any class
+   with a task, with logic that looks similar to:
+
+   ```js
+   let PATCHED = new WeakSet();
+
+   function patchWillDestroy(obj) {
+     if (PATCHED.has(obj)) return;
+
+     let oldWillDestroy = obj.willDestroy;
+
+     obj.willDestroy = function() {
+       if (oldWillDestroy) oldWillDestroy.call(this);
+
+       teardownTasks(this);
+     }
+
+     PATCHED.add(obj);
+   }
+   ```
+
+   This logic becomes especially convoluted if _multiple_ libraries are
+   attempting to patch `willDestroy` in this way.
+
+4. Finally, since this isn't a standard, it's difficult to add _layers_ of new
+   destroyable values that can interoperate with one another. For instance,
+   there is no way for `ember-concurrency` to know how to destroy tasks on
+   non-framework classes that users may have added themselves.
+
+This RFC proposes a streamlined API that disconnects the exact implementation
+from any interface, allows for multiple destructors per-destroyable, and
+maximizes interoperability in general.
+
+## Detailed design
+
+The API consists of 6 main functions, imported from `@ember/application`:
```

Comment:
FWIW, I am not sure that `@ember/application` is quite the correct layer here. A few points:

* I think it would be pretty rare to actually entangle something with the destruction of the app (the more common thing would be to associate with a service, component, route, controller, etc)
* We need to be able to use this concept in Glimmer.js applications, and having it come from `@ember/application` means that we will have a different module import in Glimmer.js than in Ember (increasing the difference in source <-> source compat between Ember and Glimmer.js). Also, it seems likely that the actual implementation of these destroyable APIs internally will come from a glimmer-vm abstraction.

Reply:
I don't think we're ready to move it to `@glimmer` just yet, especially because the ideal place would likely be `@glimmer/core`, which is still an evolving package. I think we should wait until we have an RFC for source-to-source compatibility to resolve this. Updated the RFC to `@ember/destroyable` for now, and we can move the import once we have a more thorough plan for Glimmer imports.